### PR TITLE
REST Restructure

### PR DESCRIPTION
**Pull request by AnatoleAM**
Building a nice and scalable code structure for the REST API

### CODE REVIEW

**Comment by TroyKomodo** [comment]
I will look today

File: src/server/rest/route.go [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+	URI        string
+	Method     RouteMethod
+	Children   []Route
+	Middleware []Middleware
+}
+
+type RouteMethod string
+
+const (
+	GET     RouteMethod = "GET"
```
Comment (TroyKomodo) [inline src/server/rest/route.go:24]: Does fasthttp not have http verbs as constants?

File: src/server/rest/route.go [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+type APIErrorResponse struct {
+	Status    int               `json:"status"`
+	Error     string            `json:"error"`
+	ErrorCode int               `json:"error_code"`
+	Details   map[string]string `json:"details,omitempty"`
+}
+
+type HttpStatusCode int
+
+const (
```
Comment (TroyKomodo) [inline src/server/rest/route.go:43]: I know fasthttp has status codes as constants.

File: src/server/server.go [hunk 3]
```diff
@@ -4,65 +4,95 @@ import (
...
+type HttpServer struct {
+	listener net.Listener
+	server   *fasthttp.Server
+	router   *router.Router
+	routes   map[string]*rest.Route
+}
+
+// Start: set up the http server and begin listening on the configured port
+func (s *HttpServer) Start(gCtx global.Context) (<-chan struct{}, error) {
+	var err error
```
Comment (TroyKomodo) [inline src/server/server.go:24]: I think, follow the old `New(gCtx global.Context) <-chan struct{}`  function structure.

File: src/server/server.go [hunk 4]
```diff
@@ -4,65 +4,95 @@ import (
...
-	// 404
-	app.Use(func(c *fiber.Ctx) error {
-		return c.Status(404).JSON(&fiber.Map{
-			"status":  404,
-			"message": "Not Found",
-		})
-	})
+			// Routing
+			s.router.Handler(ctx)
+		},
```
Comment (TroyKomodo) [inline src/server/server.go:69]: Add the field `CloseOnShutdown: true`

File: src/server/server.go [hunk 5]
```diff
@@ -4,65 +4,95 @@ import (
...
+		DisablePreParseMultipartForm: true,
+		LogAllErrors:                 true,
+		StreamRequestBody:            true,
+	}
 
+	// Begin listening
 	go func() {
-		err = app.Listener(ln)
-		if err != nil {
+		if err = s.server.Serve(s.listener); err != nil {
```
Comment (TroyKomodo) [inline src/server/server.go:80]: This should `defer close(done)`

File: src/server/server.go [hunk 6]
```diff
@@ -4,65 +4,95 @@ import (
...
 	}()
 
+	// Gracefully exit when the global context is canceled
 	done := make(chan struct{})
-
 	go func() {
 		<-gCtx.Done()
-		_ = app.Shutdown()
+		_ = s.server.Shutdown()
 		close(done)
```
Comment (TroyKomodo) [inline src/server/server.go:90]: Close done should be removed.

File: src/server/server.go [hunk 7]
```diff
@@ -4,65 +4,95 @@ import (
...
 	}()
 
-	return done
+	return done, err
+}
+
+func New() HttpServer {
```
Comment (TroyKomodo) [inline src/server/server.go:96]: Unneeded we dont need to expose the http server struct.

File: src/server/v3/middleware/auth.go [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+
+import (
+	"fmt"
+
+	"github.com/SevenTV/REST/src/server/rest"
+)
+
+func Auth() rest.Middleware {
+	return func(ctx *rest.Ctx) rest.APIError {
+		fmt.Println("This is middleware")
```
Comment (TroyKomodo) [inline src/server/v3/middleware/auth.go:11]: Why is this here?

File: src/server/v3/routes/auth/auth.go [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+		Method:   rest.GET,
+		Children: []rest.Route{},
+		Middleware: []rest.Middleware{
+			middleware.Auth(),
+		},
+	}
+}
+
+func (r *Route) Handler(ctx *rest.Ctx) rest.APIError {
+	fmt.Println("Auth Route")
```
Comment (TroyKomodo) [inline src/server/v3/routes/auth/auth.go:28]: Why is this here?

File: src/server/v3/routes/root.go [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+func New() rest.Route {
+	return &Route{}
+}
+
+func (r *Route) Config() rest.RouteConfig {
+	return rest.RouteConfig{
+		URI:    "/v3",
+		Method: rest.GET,
+		Children: []rest.Route{
+			auth.New(),
```
Comment (TroyKomodo) [inline src/server/v3/routes/root.go:21]: Children `auth.New()` ?

File: src/server/version.go [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+
+func (s *HttpServer) traverseRoutes(r rest.Route, parent rest.Route) {
+	c := r.Config()
+
+	// Compose the full request URI (prefixing with parent, if any)
+	uri := ""
+	if parent != nil {
+		uri = parent.Config().URI
+	}
+	uri += c.URI
```
Comment (TroyKomodo) [inline src/server/version.go:54]: This should be `uri = path.Join(uri, c.URI)`

Comment (TroyKomodo) [reply src/server/version.go:54]: Actually we shouldnt do this, we should do it with https://pkg.go.dev/github.com/fasthttp/router#Router.Group

**Review by TroyKomodo — CHANGES_REQUESTED** [review]

File: src/server/v3/middleware/auth.go (hunk 8 above)
Comment (AnatoleAM) [reply src/server/v3/middleware/auth.go:11]: It's just for debugging / testing the concept

Comment (TroyKomodo) [reply src/server/v3/middleware/auth.go:11]: If merge into dev then it should be removed.

File: src/server/rest/route.go (hunk 2 above)
Comment (AnatoleAM) [reply src/server/rest/route.go:43]: I like to have this locally, so that if the http package is ever swapped it doesn't need an update in every single route

File: src/server/v3/routes/root.go (hunk 10 above)
Comment (AnatoleAM) [reply src/server/v3/routes/root.go:21]: `auth.New()` returns a `Route`, so these are child routes. Child routes can have their own children, etc.

File: src/server/rest/route.go (hunk 2 above)
Comment (TroyKomodo) [reply src/server/rest/route.go:43]: Yes thats fine, just reference the fasthttp consts.

File: src/server/version.go (hunk 11 above)
Comment (AnatoleAM) [reply src/server/version.go]: GIGACHAD
